src/main/java/org/broadinstitute/hellbender/tools/spark/RevertSamSpark.java

```diff
@@ -0,0 +1,706 @@
+                }
+                out.add(rec);
+            }
+            return out.iterator();
+        });
+
+    }
+
+    /**
+     * Method which takes an RDD of reads that is guaranteed to have every readname group placed together on the same
+     * partition and maps those so a JavaPairRDD with the readname as the key.
+     */
+    private static JavaPairRDD<String, Iterable<GATKRead>> spanReadsByKey(final JavaRDD<GATKRead> reads) {
+        JavaPairRDD<String, GATKRead> nameReadPairs = reads.mapToPair(read -> new Tuple2<>(read.getName(), read));
+        return SparkUtils.spanByKey(nameReadPairs).flatMapToPair(namedRead -> {
+            // for each name, separate reads by key (group name)
+            List<Tuple2<String, Iterable<GATKRead>>> out = Lists.newArrayList();
+            ListMultimap<String, GATKRead> multi = LinkedListMultimap.create();
+            for (GATKRead read : namedRead._2()) {
+                multi.put(read.getName(), read);
+            }
+            for (String key : multi.keySet()) {
+                // list from Multimap is not serializable by Kryo, so put in a new array list
+                out.add(new Tuple2<>(key, Lists.newArrayList(multi.get(key))));
+            }
+            return out.iterator();
+        });
+    }
+
+    private static JavaRDD<GATKRead> querynameSortReadsIfNecessary(JavaRDD<GATKRead> reads, int numReducers, SAMFileHeader headerForTool) {
+        JavaRDD<GATKRead> sortedReadsForMarking;
+        if (ReadUtils.isReadNameGroupedBam(headerForTool)) {
+            sortedReadsForMarking = reads;
+        } else {
+            headerForTool.setSortOrder(SAMFileHeader.SortOrder.queryname);
+            sortedReadsForMarking = SparkUtils.sortReadsAccordingToHeader(reads, headerForTool, numReducers);
+        }
+        return sortedReadsForMarking;
+    }
+
+
+    private Map<String, SAMFileHeader> getReadgroupHeaderMap(SAMFileHeader inHeader, Map<String, File> writerMap) {
+        final Map<String, SAMFileHeader> headerMap;
+        if (outputByReadGroup) {
+            if (inHeader.getReadGroups().isEmpty()) {
+                throw new GATKException("The header is missing its read group map");
+            }
+
+            ValidationUtil.assertAllReadGroupsMapped(writerMap, inHeader.getReadGroups());
+            headerMap = new HashMap<>();
+            for (final SAMReadGroupRecord readGroup : inHeader.getReadGroups()) {
+                final SAMFileHeader header = createOutHeader(inHeader, sortOrder, !dontRemoveAlignmentInformation);
+                header.addReadGroup(readGroup);
+                headerMap.put(readGroup.getId(), header);
+            }
+        } else {
+            final SAMFileHeader singleOutHeader = createOutHeader(inHeader, sortOrder, !dontRemoveAlignmentInformation);
+            inHeader.getReadGroups().forEach(singleOutHeader::addReadGroup);
+            headerMap = Collections.singletonMap(null,singleOutHeader);
```

Comment:
It's super weird to have a null key in a map.  I didn't know what was even allowed in most java maps...

also:
```suggestion
            headerMap = Collections.singletonMap(null, singleOutHeader);
```